src/e2e_test/perf_tool/pkg/bq/batch_inserter.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package bq
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"cloud.google.com/go/bigquery"
+	log "github.com/sirupsen/logrus"
+)
+
+// BatchInserter batches up bigquery struct savers, to reduce the number of calls to inserter.Put.
+type BatchInserter struct {
+	// BatchSize specifies how many rows (struct savers) to wait for before pushing a batch.
+	BatchSize int
+	// PushTimeout specifies how long to wait before pushing a batch even if the batch isn't at BatchSize yet.
+	PushTimeout time.Duration
+	Table       *Table
+
+	batch     []*bigquery.StructSaver
+	batchLock sync.Mutex
+	batchSize atomic.Int64
+	inserter  *bigquery.Inserter
+}
+
+// Run starts running the BatchInserter.
+func (b *BatchInserter) Run(rowCh <-chan interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	b.inserter = b.Table.Inserter()
+	b.inserter.SkipInvalidRows = false
+
+	b.batchLock.Lock()
+	b.batch = make([]*bigquery.StructSaver, 0)
+	b.batchSize.Store(0)
+	b.batchLock.Unlock()
+
+	go b.batchWatchdog(ctx)
+	for row := range rowCh {
+		s := &bigquery.StructSaver{
+			Struct: row,
+			Schema: b.Table.Schema,
+		}
+		b.batchLock.Lock()
+		b.batch = append(b.batch, s)
+		b.batchSize.Add(1)
+		b.batchLock.Unlock()
+	}
+}
+
+func (b *BatchInserter) batchWatchdog(ctx context.Context) {
+	t := time.NewTimer(b.PushTimeout)
+	for {
+		select {
+		case <-ctx.Done():
+			b.uploadBatch()
+			return
+		case <-t.C:
+			b.uploadBatch()
+			t.Reset(b.PushTimeout)
+		default:
+			if b.batchSize.Load() >= int64(b.BatchSize) {
+				b.uploadBatch()
+				if !t.Stop() {
+					<-t.C
+				}
```

Comment:
I think you can remove the need for mutex-es and a separate watchdog by combining these.

something like:
```
func (b* BatchInserter) Run(data <- ch interface{}) {
  for {
    select {
    case <-t.C:
      b.Upload()
    case datum := <-data:
      if len(batch) >= maxLen:
        b.Upload()
      b.Add(datum)
    }
  }
}
```